Retry round tripper

SimonRichardson opened the pull request.

SimonRichardson committed:
Middleware: Introduce a retry roundtripper

This is the beginnings of the retry transport. The idea being that we
allow global retries depending on the errors passed in.

This isn't smart about recommencing already downloaded content, this is
too low down to be able to do that, instead this is a generic attempt at
solving the problem where network stacks are flakey.

SimonRichardson committed:
Middleware: Add testing for the retry middleware

SimonRichardson committed:
Client: Wire up retry policy

The following adds the retry setup along with tests for seeing the retry
policy working.

SimonRichardson committed:
Retry: Test back-off

The following ensures that we read the Retry-After header to calculate
the retry backoff.

SimonRichardson commented:
!!build!!

hmlanigan commented on middleware.go:
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
+}
+
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
```
"The "?

hmlanigan commented on middleware.go:
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
+	return "retryable error"
+}
+
+func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
```
Method comment header needed.

hmlanigan commented on middleware.go:
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
 }
+
+// RetryMiddleware allows retrying of certain retryable http errors.
+// This only handles very specific status codes, ones that are deemed retryable;
```
s/;/:/

hmlanigan commented on middleware.go:
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
+func (m retryMiddleware) defaultBackoff(resp *http.Response, max time.Duration) time.Duration {
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusTooManyRequests {
```
The RFC indicated that only 503 may return Retry-After, doesn't hurt to check others.

hmlanigan commented on middleware.go:
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusTooManyRequests {
+		if header := resp.Header.Get("Retry-After"); header != "" {
+			// Attempt to parse the header from the request.
+			retry, err := strconv.ParseInt(header, 10, 64)
```
Found the hard way, the retry value can also be in time.RFC1123 format: https://github.com/go-goose/goose/blob/v2/http/client.go#L384-L409

SimonRichardson replied:
NICE, thanks for this. I wonder if we can maybe drop this one in goose eventually as we can pass in a juju/http client? Either way this is great information.

Fixed in e47bd4b

SimonRichardson replied:
https://github.com/juju/http/pull/6/files#diff-02f50901140d6057da6310a106670552aa766a093efbc2200fb34c099b762131R200-R202

SimonRichardson committed:
Middleware: handle retries with dates

According to the RFC7231 spec it states that the Retry-After header
could actually be either a delay in seconds or a date in the future. We
only previously handled delays in seconds and that would have failed
because we just didn't correctly handle it.

The fix was to handle that and also so clamp it to a maximum duration as
we would never want to wait for too long as that could potentially dead
lock the running program.

SimonRichardson commented:
!!build!!

SimonRichardson commented:
!!build!!

hmlanigan commented on middleware.go:
```diff
@@ -235,17 +248,36 @@ func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, err
+		date, err := time.Parse(time.RFC1123, header)
+		if err == nil {
+			return m.clampBackoff(m.clock.Now().Sub(date))
 		}
```
We should log bad header data, that the Retry-After value is not valid, and what the value is.

hmlanigan commented on middleware.go:
```diff
@@ -235,17 +248,36 @@ func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, err
+	if header := resp.Header.Get("Retry-After"); header != "" {
+		// Attempt to parse the header from the request.
+		//
+		// Check for delay in seconds first, before checkking for a http-date
```
s/checkking/checking/

hmlanigan commented on middleware.go:
```diff
@@ -207,7 +218,9 @@ func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
 		BackoffFunc: func(delay time.Duration, attempts int) time.Duration {
-			return m.defaultBackoff(res, delay)
+			var duration time.Duration
+			duration, backOffErr = m.defaultBackoff(res, delay)
```
where is backOffErr evaluated? or logged?

SimonRichardson replied:
https://github.com/juju/http/pull/6/files#diff-02f50901140d6057da6310a106670552aa766a093efbc2200fb34c099b762131R200-R202

hmlanigan reviewed (approved):
A few minor things to resolve, o.w. LGTM

SimonRichardson committed:
Retry: Log when backoff fails

SimonRichardson commented:
$$merge$$
